lib/improver/standardise.py

```diff
@@ -0,0 +1,442 @@
+from improver.metadata.amend import amend_attributes
+from improver.metadata.check_datatypes import (
+    check_cube_not_float64, check_time_coordinate_metadata)
+from improver.metadata.constants.time_types import (
+    TIME_COORD_NAMES, TIME_REFERENCE_DTYPE, TIME_REFERENCE_UNIT,
+    TIME_DTYPE, TIME_UNIT)
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_checker import spatial_coords_match
+from improver.utilities.spatial import OccurrenceWithinVicinity
+
+
+def grid_contains_cutout(grid, cutout):
+    """
+    Check that a spatial cutout is contained within a given grid
+
+    Args:
+        grid (iris.cube.Cube):
+            A cube defining a data grid
+        cutout (iris.cube.Cube):
+            The cutout to search for within the grid
+
+    Returns:
+        bool:
+            True if cutout is contained within grid, False otherwise
+    """
+    if spatial_coords_match(grid, cutout):
+        return True
+
+    # check whether "cutout" coordinate points match a subset of "grid"
+    # points on both axes
+    for axis in ['x', 'y']:
+        grid_coord = grid.coord(axis=axis)
+        cutout_coord = cutout.coord(axis=axis)
+        # check coordinate metadata
+        if (cutout_coord.name() != grid_coord.name() or
+                cutout_coord.units != grid_coord.units or
+                cutout_coord.coord_system != grid_coord.coord_system):
+            return False
+
+        # search for cutout coordinate points in larger grid
+        cutout_start = cutout_coord.points[0]
+        if not np.any([np.isclose(cutout_start, grid_point)
+                       for grid_point in grid_coord.points]):
+            return False
+
+        start = list(grid_coord.points).index(cutout_start)
+        end = start + len(cutout_coord.points)
+        if not np.allclose(cutout_coord.points, grid_coord.points[start:end]):
+            return False
+
+    return True
+
+
+class StandardiseGridAndMetadata(BasePlugin):
+
+    """Plugin to regrid cube data and standardise metadata"""
+
+    REGRID_REQUIRES_LANDMASK = {"bilinear": False,
+                                "nearest": False,
+                                "nearest-with-mask": True}
+
+    def __init__(self, regrid_mode='bilinear', extrapolation_mode='nanmask',
+                 landmask=None, landmask_vicinity=25000, grid_attributes=None):
+        """
+        Initialise regridding parameters
+
+        Args:
+            regrid_mode (str):
+                Mode of interpolation in regridding.
+            extrapolation_mode (str):
+                Mode to fill regions outside the domain in regridding.
+            landmask (iris.cube.Cube or None):
+                Land-sea mask ("land_binary_mask") on the input cube grid.
+                Required for "nearest-with-mask" regridding option.
+            landmask_vicinity (float):
+                Radius of vicinity to search for a coastline, in metres
+            grid_attributes (list of str or None):
```

Comment:
No unit-test uses this argument.

Reply:
Good point.  Added a couple of new tests around this argument, explicitly testing the default and configurable behaviour separately from the attribute updates test.